extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java

```diff
@@ -367,19 +397,26 @@ private void createFrameWriterIfNeeded(RowsAndColumns rac, AtomicInteger rowId)
   public void writeRacToFrame(RowsAndColumns rac, AtomicInteger rowId) throws IOException
   {
     final int numRows = rac.numRows();
-    rowId.set(0);
     while (rowId.get() < numRows) {
-      final boolean didAddToFrame = frameWriter.addSelection();
-      if (didAddToFrame) {
+      if (frameWriter.addSelection()) {
+        incrementBoostColumn();
         rowId.incrementAndGet();
-        partitionBoostVirtualColumn.setValue(partitionBoostVirtualColumn.getValue() + 1);
-      } else if (frameWriter.getNumRows() == 0) {
-        throw new FrameRowTooLargeException(currentAllocatorCapacity);
-      } else {
+      } else if (frameWriter.getNumRows() > 0) {
         flushFrameWriter();
-        return;
```

Comment:
the fix of the bug should be to call `createFrameWriterIfNeeded(rac, rowId);` instead of the `return` here; all other changes are not necessary.

I think the `testFrameWriterReachingCapacity` is not fair; I don't think there is a contract anywhere that a processor may only write a single output during a single call

Reply:
> I think the testFrameWriterReachingCapacity is not fair; I don't think there is a contract anywhere that a processor may only write a single output during a single call

[runIncrementally's contract](https://github.com/apache/druid/blob/3d56fa6f5607d9b56f6da7705f4fb2ec8b6803ce/processing/src/main/java/org/apache/druid/frame/processor/FrameProcessor.java#L54) states that:
```
This method must not read more than one frame from each readable input channel, and must not write more than one frame to each output channel.
```

> the fix of the bug should be to call createFrameWriterIfNeeded(rac, rowId); instead of the return here; all other changes are not necessary.

It doesn't work. It errors out with `Channel has no capacity`.

Gian also had to fix the same issue for ScanQueryFrameProcessor in https://github.com/apache/druid/pull/13036.

Laksh also pointed out the same thing in an older PR of mine: https://github.com/apache/druid/pull/16823#discussion_r1706972367